docs/guides/dropout.rst

```diff
@@ -0,0 +1,147 @@
+:meth:`flax.linen.Module` is the
+`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__
+, and all layers and models are subclassed from it.
+
+In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is ``None`` by
+default.
+
+* If it's set to ``False``, the inputs are scaled by ``1 / (1 - dropout_rate)`` and masked.
+* When it's ``True``, no mask is applied (the dropout is turned off).
+
+Flax provides an implicit way of handling PRNG key streams via Flax
+:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
+It allows the code in Flax ``Module``s (or their sub-``Module``s) to
+"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
+call it.
+
+**Note:** Internally, :meth:`flax.linen.Dropout` makes use of :meth:`flax.linen.Module.make_rng`
+to create a key for dropout. You can check out the
+`source code <https://github.com/google/flax/blob/5714e57a0dc8146eb58a7a06ed768ed3a17672f9/flax/linen/stochastic.py#L72>`__.
+
+* Every time :meth:`Module.make_rng` is called (which is done implicitly in
+  :meth:`flax.linen.Dropout`), you get a new PRNG key split from the main/root pseudorandom
+  number generator (PRNG) key.
+* :meth:`Module.make_rng` guarantees full reproducibility.
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+
+
+
+      return x
+
+  ---
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+      # Set the dropout layer with a rate of 50%. #!
+      # When the `deterministic` flag is `True`, dropout is turned off. #!
+      x = nn.Dropout(rate=0.5, deterministic=not self.training)(x) #!
+      return x
+
+After adding ``Dropout`` to your model, instantiate the model with
+``training=True``.
+
+For example, when creating a ``TrainState``, the code will include::
+
+  my_model = MyModel(num_neurons=3, training=True)
+
+Initializing the model
+**********************
+
+The PRNG key provided to :meth:`flax.linen.init()` is the one that seeds the
+``'params'`` PRNG key stream.
+
+Initialize the model with the ``params'`` key::
+
+  x = jax.random.uniform(key=main_key, shape=(3, 4, 4))
+  variables = my_model.init(params_key, x)
+
+Performing the forward pass during training
+*******************************************
+
+To draw PRNG keys during the forward pass (with dropout), provide a PRNG key to
+seed that ``'dropout'`` stream when you call :method:`flax.linen.apply()`.
+
+In your train step, the code will include:
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  y = my_model.apply(variables, x)
+
+  ---
+  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```

Comment:
I would clarify here that you should construct the Module here with the right keyword arguments.

```suggestion
  my_model = MyModel(num_neurons=3, training=False)
  y = my_model.apply(variables, x)

  ---
  my_model = MyModel(num_neurons=3, training=True)
  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```

Reply:
Good idea.

```
.. codediff::
  :title_left: Dropout disabled
  :title_right: Dropout enabled
  :sync:

  my_model_without_dropout = MyModel(num_neurons=3, training=False)
  y = my_model.apply(variables, x)

  ---
  my_model_with_dropout = MyModel(num_neurons=3, training=True) #!
  y = my_model.apply(variables, x, rngs={'dropout': dropout_key}) #!
```